src/cnf.cpp

```diff
@@ -0,0 +1,371 @@
+    // minimum and maximum variable defined in the clause
+    unsigned min_var = vars.size(), max_var = 0;
+    // whether the current clause is a tautology
+    bool is_tautology = false;
+
+    for (int literal : clauses_raw) {
+      if (is_tautology) {
+        if (literal == 0) {
+          is_tautology = false;
+          std::fill(polarities.begin() + min_var, polarities.begin() + max_var + 1, 0);
+        }
+        continue;
+      }
+
+      if (literal != 0) {
+        const unsigned var         = vars[std::abs(literal) - 1];
+        const signed char polarity = literal < 0 ? -1 : 1, prev = polarities[var];
+        if (prev != 0 && prev != polarity) {
+          is_tautology = true; // x ∨ ¬x ≡ ⊤
+        } else {
+          polarities[var] = polarity;
+          if (var > max_var) max_var = var;
+          if (var < min_var) min_var = var;
+        }
+        continue;
+      }
+
+      if (min_var == vars.size()) // ⊥
+        return { adapter.bot() };
+
+      unsigned level = vars.size() - 1;
+
+      typename Adapter::build_node_t tautology = adapter.build_node(true);
+      for (; level != max_var; --level) {
+        tautology = adapter.build_node(level, tautology, tautology);
+      }
+
+      assert(level == max_var);
+      assert(polarities[level] == 1 || polarities[level] == -1);
+      typename Adapter::build_node_t clause_build = polarities[level] == 1
+        ? adapter.build_node(level, /* lo */ adapter.build_node(false), tautology)
+        : tautology;
+      if (level > min_var) tautology = adapter.build_node(level, tautology, tautology);
+      polarities[level] = 0;
+
+      while (level-- != 0) {
+        const signed char pol = polarities[level];
+        if (pol == 0) {
+          clause_build = adapter.build_node(level, clause_build, clause_build);
+        } else {
+          polarities[level] = 0;
+          clause_build      = pol == 1 ? adapter.build_node(level, /* lo */ clause_build, tautology)
+                                       : adapter.build_node(level, /* lo */ tautology, clause_build);
+        }
+        if (level > min_var) tautology = adapter.build_node(level, tautology, tautology);
+      }
+
+      clauses.push_back(adapter.build());
+      min_var = vars.size();
+      max_var = 0;
+    }
+  }
+
+  return clauses;
+}
+
+/// Conjoin the clauses with an (approximately) balanced bracketing `(c1 ∧ c2) ∧ (c3 ∧ c4)`
+///
+/// In the paper [Configuring BDD Compilation Techniques for Feature
+/// Models](https://doi.org/10.1145/3646548.3676538), this was shown to be much
+/// better than a left-deep clause bracketing `((c1 ∧ c2) ∧ c3) ∧ c4`.
+template <typename Adapter, typename IT>
+typename Adapter::dd_t
+conjoin(Adapter& adapter, const IT begin, const IT end)
+{
+  if (begin == end) return adapter.top();
+  auto d = std::distance(begin, end);
+  if (d == 1) return *begin;
+
+  const IT mid = begin + d / 2;
+  return conjoin(adapter, begin, mid) & conjoin(adapter, mid, end);
+}
```

Comment:
The advantage of the recursive function is that it is easy to read. There is no need to worry about the recursion depth, since it will be at most `O(log(#clauses))` (or less than 64 in practice). Moreover, we would like to preserve the input clause ordering, since it is a good idea to process clauses sharing a lot of variables together. If I see it correctly, your algorithm would violate this. If we had `1 ∧ 2 ∧ 3 ∧ 4 ∧ 5`, then it would compute `(5 ∧ (1 ∧ 2)) ∧ (3 ∧ 4)` instead of `(1 ∧ 2) ∧ ((3 ∧ 4) ∧ 5)`.

Reply:
That is a very valid point. I suppose, there is right now an implicit assumption that clauses are given in some good order?

This might be left as future work (or even better a separate strategy that can be chosen with some flag):
- Try to compute the "distance" between clauses to then sort them into buckets. One can maybe take inspiration from Locality-sensitive hashing .
- And maybe also just go all the way to add *Bucket Elimination* from the paper by Randy you correctly identified.

But fair, let's add this as a TODO comment and mark this as *resolved*.